Add logic in CSI manager to return FaultType when calling CNS API.

lipingxue opened the pull request:
Add logic for manger to return FaultType for CreateVolume, DeleteVolume, AttachVolume, DetachVolume and ExpandVolume.

Also add logic to return the fault type from manager back to controller layer. With the change,  at controller layer,
we know the FaultType returned when calling CNS APIs.

<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Testing done**:
E2E test for Block Vanilla, File Vanilla, WCP and GC all Passed.

The following manual testing has been done.

Test 1: Attach Volume failed by throwing NotFound

1. create a pvc, and make sure the pvc is bound
2. delete the fcd disk corresponding to the pv
3. create a pod to use this pvc
4. observed Attach Volume failed with NotFound
5. Check the CSI log observed the NotFound fault is thrown by CNS AttachVolume API. This fault is captured by manager AttachVolume API and also populated to the  controllerPublishVolumeInternal at controller layer.

```
2021-07-29T23:36:00.741Z        ERROR   volume/manager.go:484   CNS AttachVolume failed from vCenter "sc-rdops-vm14-dhcp-165-12.eng.vmware.com" with err: ServerFaultCode: CNS: Failed to retrieve datastore for vol 312002bb-dbb1-438b-a4e4-951978282946. (vim.fault.NotFound) {
   faultCause = (vmodl.MethodFault) null,
   faultMessage = <unset>
   msg = "The vStorageObject (vim.vslm.ID) {
   dynamicType = null,
   dynamicProperty = null,
   id = 312002bb-dbb1-438b-a4e4-951978282946
} was not found"
}       {"TraceId": "65043a8b-1300-424c-a59b-6279c45545f1"}
sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/volume.(*defaultManager).AttachVolume.func1
        /build/pkg/common/cns-lib/volume/manager.go:484
sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/volume.(*defaultManager).AttachVolume
        /build/pkg/common/cns-lib/volume/manager.go:539
sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/common.AttachVolumeUtil
        /build/pkg/csi/service/common/vsphereutil.go:456
sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/wcp.(*controller).ControllerPublishVolume.func1
        /build/pkg/csi/service/wcp/controller.go:662
sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/wcp.(*controller).ControllerPublishVolume
        /build/pkg/csi/service/wcp/controller.go:712
github.com/container-storage-interface/spec/lib/go/csi._Controller_ControllerPublishVolume_Handler.func1
        /go/pkg/mod/github.com/container-storage-interface/spec@v1.4.0/lib/go/csi/csi.pb.go:5632
github.com/rexray/gocsi/middleware/serialvolume.(*interceptor).controllerPublishVolume
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware/serialvolume/serial_volume_locker.go:120
github.com/rexray/gocsi/middleware/serialvolume.(*interceptor).handle
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware/serialvolume/serial_volume_locker.go:86
github.com/rexray/gocsi/utils.ChainUnaryServer.func2.1.1
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/utils/utils_middleware.go:99
github.com/rexray/gocsi/middleware/specvalidator.(*interceptor).handleServer.func1
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware/specvalidator/spec_validator.go:178
github.com/rexray/gocsi/middleware/specvalidator.(*interceptor).handle
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware/specvalidator/spec_validator.go:218
github.com/rexray/gocsi/middleware/specvalidator.(*interceptor).handleServer
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware/specvalidator/spec_validator.go:177
github.com/rexray/gocsi/utils.ChainUnaryServer.func2.1.1
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/utils/utils_middleware.go:99
github.com/rexray/gocsi.(*StoragePlugin).injectContext
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware.go:231
github.com/rexray/gocsi/utils.ChainUnaryServer.func2.1.1
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/utils/utils_middleware.go:99
github.com/rexray/gocsi/utils.ChainUnaryServer.func2
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/utils/utils_middleware.go:106
github.com/container-storage-interface/spec/lib/go/csi._Controller_ControllerPublishVolume_Handler
        /go/pkg/mod/github.com/container-storage-interface/spec@v1.4.0/lib/go/csi/csi.pb.go:5634
google.golang.org/grpc.(*Server).processUnaryRPC
        /go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1024
google.golang.org/grpc.(*Server).handleStream
        /go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1313
google.golang.org/grpc.(*Server).serveStreams.func1.1
        /go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:722
 
 
 
 
2021-07-29T23:36:00.741Z        INFO    vsphere/utils.go:100    Extract vimfault type: +types.NotFound. SoapFault Info: +&{{http://schemas.xmlsoap.org/soap/envelope/ Fault} ServerFaultCode CNS: Failed to retrieve datastore for vol 312002bb-dbb1-438b-a4e4-951978282946. (vim.fault.NotFound) {
   faultCause = (vmodl.MethodFault) null,
   faultMessage = <unset>
   msg = "The vStorageObject (vim.vslm.ID) {
   dynamicType = null,
   dynamicProperty = null,
   id = 312002bb-dbb1-438b-a4e4-951978282946
} was not found"
} {{{{<nil> []}}}}} from err +ServerFaultCode: CNS: Failed to retrieve datastore for vol 312002bb-dbb1-438b-a4e4-951978282946. (vim.fault.NotFound) {
   faultCause = (vmodl.MethodFault) null,
   faultMessage = <unset>
   msg = "The vStorageObject (vim.vslm.ID) {
   dynamicType = null,
   dynamicProperty = null,
   id = 312002bb-dbb1-438b-a4e4-951978282946
} was not found"
}       {"TraceId": "65043a8b-1300-424c-a59b-6279c45545f1"}
2021-07-29T23:36:00.741Z        DEBUG   volume/manager.go:540   internalAttachVolume: returns fault "vim.fault.NotFound" for volume "312002bb-dbb1-438b-a4e4-951978282946"      {"TraceId": "65043a8b-1300-424c-a59b-6279c45545f1"}
 
2021-07-29T23:36:00.741Z        DEBUG   wcp/controller.go:698   AttachVolumeUtil returns fault vim.fault.NotFound:      {"TraceId": "65043a8b-1300-424c-a59b-6279c45545f1"}
 
 
2021-07-29T23:36:00.741Z        DEBUG   wcp/controller.go:714   controllerPublishVolumeInternal: returns fault "vim.fault.NotFound" for volume "312002bb-dbb1-438b-a4e4-951978282946"   {"TraceId": "65043a8b-1300-424c-a59b-6279c45545f1"}
```

Test 2: Delete Volume failed with ResourceInUse

1. create a pvc and wait for pvc bound
2. Use VC UI to attach the volume to a dummy VM. You can figure out which fcd to attach with the create timestamp on fcd.
3. delete the pvc
4. observed pvc delete failed with ResourceInUse
5.  Checked the CSI log, in the task returned from CNS DeleteVolume API, it has fault "ResourceInUse". This fault is captured by manager DeleteVolume API and also populated to the deleteVolumeInternal at controller layer.
```
2021-07-29T23:32:09.840Z        ERROR   volume/manager.go:738   failed to delete volume: "1ead9ebd-f418-4473-85da-f02fc5410561", fault: "(*types.LocalizedMethodFault)(0xc00053aee0)({\n DynamicData: (types.DynamicData) {\n },\n Fault: (*types.ResourceInUse)(0xc0007110c0)({\n  VimFault: (types.VimFault) {\n   MethodFault: (types.MethodFault) {\n    FaultCause: (*types.LocalizedMethodFault)(<nil>),\n    FaultMessage: ([]types.LocalizableMessage) <nil>\n   }\n  },\n  Type: (string) \"\",\n  Name: (string) (len=6) \"volume\"\n }),\n LocalizedMessage: (string) (len=32) \"The resource 'volume' is in use.\"\n})\n", opID: "b3627de2" {"TraceId": "9351fa0c-674f-4779-b296-2dd1060ef9ce"}
sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/volume.(*defaultManager).deleteVolume
        /build/pkg/common/cns-lib/volume/manager.go:738
sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/volume.(*defaultManager).DeleteVolume.func1
        /build/pkg/common/cns-lib/volume/manager.go:678
sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/volume.(*defaultManager).DeleteVolume
        /build/pkg/common/cns-lib/volume/manager.go:681
sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/common.DeleteVolumeUtil
        /build/pkg/csi/service/common/vsphereutil.go:489
sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/wcp.(*controller).DeleteVolume.func1
        /build/pkg/csi/service/wcp/controller.go:576
sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/wcp.(*controller).DeleteVolume
        /build/pkg/csi/service/wcp/controller.go:584
github.com/container-storage-interface/spec/lib/go/csi._Controller_DeleteVolume_Handler.func1
        /go/pkg/mod/github.com/container-storage-interface/spec@v1.4.0/lib/go/csi/csi.pb.go:5614
github.com/rexray/gocsi/middleware/serialvolume.(*interceptor).deleteVolume
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware/serialvolume/serial_volume_locker.go:183
github.com/rexray/gocsi/middleware/serialvolume.(*interceptor).handle
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware/serialvolume/serial_volume_locker.go:92
github.com/rexray/gocsi/utils.ChainUnaryServer.func2.1.1
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/utils/utils_middleware.go:99
github.com/rexray/gocsi/middleware/specvalidator.(*interceptor).handleServer.func1
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware/specvalidator/spec_validator.go:178
github.com/rexray/gocsi/middleware/specvalidator.(*interceptor).handle
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware/specvalidator/spec_validator.go:218
github.com/rexray/gocsi/middleware/specvalidator.(*interceptor).handleServer
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware/specvalidator/spec_validator.go:177
github.com/rexray/gocsi/utils.ChainUnaryServer.func2.1.1
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/utils/utils_middleware.go:99
github.com/rexray/gocsi.(*StoragePlugin).injectContext
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/middleware.go:231
github.com/rexray/gocsi/utils.ChainUnaryServer.func2.1.1
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/utils/utils_middleware.go:99
github.com/rexray/gocsi/utils.ChainUnaryServer.func2
        /go/pkg/mod/github.com/rexray/gocsi@v1.2.2/utils/utils_middleware.go:106
github.com/container-storage-interface/spec/lib/go/csi._Controller_DeleteVolume_Handler
        /go/pkg/mod/github.com/container-storage-interface/spec@v1.4.0/lib/go/csi/csi.pb.go:5616
google.golang.org/grpc.(*Server).processUnaryRPC
        /go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1024
google.golang.org/grpc.(*Server).handleStream
        /go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1313
google.golang.org/grpc.(*Server).serveStreams.func1.1
        /go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:722
 
 
 
 
2021-07-29T23:32:09.841Z        DEBUG   volume/manager.go:682   internalDeleteVolume: returns fault "vim.fault.ResourceInUse" for volume "1ead9ebd-f418-4473-85da-f02fc5410561" {"TraceId": "9351fa0c-674f-4779-b296-2dd1060ef9ce"}
 
 
2021-07-29T23:32:09.842Z        DEBUG   wcp/controller.go:578   DeleteVolumeUtil returns fault vim.fault.ResourceInUse: {"TraceId": "9351fa0c-674f-4779-b296-2dd1060ef9ce"}

2021-07-29T23:32:09.843Z        DEBUG   wcp/controller.go:586   deleteVolumeInternal: returns fault "vim.fault.ResourceInUse" for volume "1ead9ebd-f418-4473-85da-f02fc5410561" {"TraceId": "9351fa0c-674f-4779-b296-2dd1060ef9ce"}
```


**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

svcbot-qecnsdp commented:
Started Vanilla file pre-checkin pipeline... Build Number: 71

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 122

svcbot-qecnsdp commented:
``` 
Build ID: 122
WCP build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 13 of 214 Specs in 2644.406 seconds
SUCCESS! -- 13 Passed | 0 Failed | 0 Pending | 201 Skipped
PASS

Ginkgo ran 1 suite in 44m41.434948471s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/csi-wcp-pre-check-in/Results/122/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
``` 
Build ID: 71
File vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
Ran 25 of 214 Specs in 8806.365 seconds
FAIL! -- 18 Passed | 7 Failed | 0 Pending | 189 Skipped
--- FAIL: TestE2E (8806.44s)
FAIL

Ginkgo ran 1 suite in 2h27m29.364102124s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-file-vanilla-pre-check-in@2/71/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started Vanilla file pre-checkin pipeline... Build Number: 73

svcbot-qecnsdp commented:
``` 
Build ID: 73
File vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
Ran 25 of 214 Specs in 8776.574 seconds
FAIL! -- 18 Passed | 7 Failed | 0 Pending | 189 Skipped
--- FAIL: TestE2E (8776.65s)
FAIL

Ginkgo ran 1 suite in 2h26m50.246824806s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-file-vanilla-pre-check-in/73/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started Vanilla file pre-checkin pipeline... Build Number: 76

svcbot-qecnsdp commented:
``` 
Build ID: 76
File vanilla build status: SUCCESS 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 

Ran 25 of 214 Specs in 5294.463 seconds
SUCCESS! -- 25 Passed | 0 Failed | 0 Pending | 189 Skipped
PASS

Ginkgo ran 1 suite in 1h28m53.222132373s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/csi-file-vanilla-pre-check-in/76/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started Vanilla block pre-checkin pipeline... Build Number: 199

svcbot-qecnsdp commented:
``` 
Build ID: 199
Block vanilla build status: SUCCESS 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
JUnit report was created: /home/worker/workspace/csi-block-vanilla-pre-check-in/199/vsphere-csi-driver/tests/e2e/junit.xml

Ran 1 of 214 Specs in 302.728 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 213 Skipped
PASS

Ginkgo ran 1 suite in 5m39.0089759s
Test Suite Passed
--
JUnit report was created: /home/worker/workspace/csi-block-vanilla-pre-check-in/199/vsphere-csi-driver/tests/e2e/junit.xml

Ran 8 of 214 Specs in 3204.687 seconds
SUCCESS! -- 8 Passed | 0 Failed | 0 Pending | 206 Skipped
PASS

Ginkgo ran 1 suite in 53m36.883495213s
Test Suite Passed
--
------------------------------

Ran 38 of 214 Specs in 657.657 seconds
SUCCESS! -- 38 Passed | 0 Failed | 0 Pending | 176 Skipped


Ginkgo ran 1 suite in 11m12.2501329s
Test Suite Passed

```

svcbot-qecnsdp commented:
Started GC block pre-checkin pipeline... Build Number: 77

svcbot-qecnsdp commented:
``` 
Build ID: 77
GC build status: FAILURE 
Stage before exit: checkout 

```

svcbot-qecnsdp commented:
Started GC block pre-checkin pipeline... Build Number: 78

svcbot-qecnsdp commented:
``` 
Build ID: 78
GC build status: FAILURE 
Stage before exit: checkout 

```

svcbot-qecnsdp commented:
Started GC block pre-checkin pipeline... Build Number: 81

svcbot-qecnsdp commented:
``` 
Build ID: 81
GC build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 1 of 214 Specs in 74.848 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 213 Skipped
PASS

Ginkgo ran 1 suite in 1m50.954261348s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/gc-csi-pre-checkin/Results/81/vsphere-csi-driver`

```

deepakkinni commented on pkg/common/cns-lib/volume/manager.go:
```diff
@@ -54,19 +54,34 @@ const (
 type Manager interface {
 	// CreateVolume creates a new volume given its spec.
-	CreateVolume(ctx context.Context, spec *cnstypes.CnsVolumeCreateSpec) (*CnsVolumeInfo, error)
+	CreateVolume(ctx context.Context, spec *cnstypes.CnsVolumeCreateSpec) (*CnsVolumeInfo, string, error)
```
This is not intuitive at all, one cannot tell what the string is by just looking at the interface. Furthermore, there is a string and error, both of which provide the same information to the caller, i.e, there was an issue when invoking the API. I think creating a struct with {error, faultType} and returning that instead of a plain error is a better approach.

chethanv28 replied:
+1

deepakkinni reviewed (commented):
Can you fill in the description as to why this change is needed. thanks.

chethanv28 commented on pkg/common/cns-lib/volume/manager.go:
```diff
@@ -54,19 +54,34 @@ const (
+	CSIInternalFault = "csi.fault.Internal"
+	// CSINotFoundFault is the fault type returned when object required is not found.
+	CSINotFoundFault = "csi.fault.NotFound"
+	// CNSInvalidArgumentFault is the fault type returned when invalid argument is given.
```
`CSIInvalidArgumentFault`

chethanv28 commented on pkg/common/cns-lib/volume/manager.go:
```diff
@@ -54,19 +54,34 @@ const (
 	taskInvocationStatusError      = cnsvolumeoperationrequest.TaskInvocationStatusError
+
+	// CSITaskInfoEmptyFault is the fault type when taskInfo is empty.
+	CSITaskInfoEmptyFault = "csi.fault.TaskInfoEmpty"
```
I don't think the error strings should be defined here. They are prefixed with CSI & the package is CNS-lib. IMO, they don't belong to this package

gohilankit commented:
/approve

lipingxue commented:
> Can you take care of Deepak's comment regarding error construct for API interface, or discuss it with the team once?

I have added comments in the interface definition to point out the requirement for the return value for the interfaces I modified. Define a new struct to encapsulate faultType and error is over complex. @gohilankit

gohilankit commented:
/approve

chethanv28 reviewed (commented):
/ok-to-test

svcbot-qecnsdp commented:
Started Vanilla block pre-checkin pipeline... Build Number: 317

svcbot-qecnsdp commented:
``` 
Build ID: 317
Block vanilla build status: SUCCESS 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
JUnit report was created: /home/worker/workspace/csi-block-vanilla-pre-check-in/317/vsphere-csi-driver/tests/e2e/junit.xml

Ran 7 of 226 Specs in 1136.347 seconds
SUCCESS! -- 7 Passed | 0 Failed | 0 Pending | 219 Skipped
PASS

Ginkgo ran 1 suite in 19m36.0004173s
Test Suite Passed

```

svcbot-qecnsdp commented:
Started Vanilla block pre-checkin pipeline... Build Number: 318

svcbot-qecnsdp commented:
Started Vanilla block pre-checkin pipeline... Build Number: 321

svcbot-qecnsdp commented:
``` 
Build ID: 321
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 

Ran 4 of 226 Specs in 910.234 seconds
FAIL! -- 2 Passed | 2 Failed | 0 Pending | 222 Skipped
--- FAIL: TestE2E (910.33s)
FAIL

Ginkgo ran 1 suite in 15m57.840999269s
Test Suite Failed

```

svcbot-qecnsdp commented:
Started Vanilla block pre-checkin pipeline... Build Number: 322

svcbot-qecnsdp commented:
``` 
Build ID: 322
Block vanilla build status: SUCCESS 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
JUnit report was created: /home/worker/workspace/csi-block-vanilla-pre-check-in/322/vsphere-csi-driver/tests/e2e/junit.xml

Ran 1 of 226 Specs in 224.283 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 225 Skipped
PASS

Ginkgo ran 1 suite in 4m35.655416146s
Test Suite Passed

```

chethanv28 commented:
/lgtm
